test/variable-cpp.stderr

```diff
@@ -0,0 +1,3 @@
+ERROR: variable.c:34: default initialization of an object of const type 'const char *(*const)(const char *)'
+ERROR: variable.c:39: default initialization of an object of const type 'const char *(*const volatile)(const char *)'
+ERROR: variable.c:44: default initialization of an object of const type 'const char *(*const *volatile *const[12][3])(const char *)'
```

Comment:
Here's something weird. I need to do this to make the tests pass:

```
diff --git a/test/variable-cpp.stderr b/test/variable-cpp.stderr
index 91db546cb779..681c232bf628 100644
--- a/test/variable-cpp.stderr
+++ b/test/variable-cpp.stderr
@@ -1,3 +1,3 @@
 ERROR: variable.c:34: default initialization of an object of const type 'const char *(*const)(const char *)'
 ERROR: variable.c:39: default initialization of an object of const type 'const char *(*const volatile)(const char *)'
-ERROR: variable.c:44: default initialization of an object of const type 'const char *(*const *volatile *const[12][3])(const char *)'
+ERROR: variable.c:44: default initialization of an object of const type 'const char *(*const *volatile *const [12][3])(const char *)'
```

i.e. add space.

I assume this works for you and looks like in github actions as well.

Clang 11 here.

Reply:
Clang 14 in github actions:

```
The following additional packages will be installed:
  libclang-14-dev python3-clang-14
```

Reply:
> something to consider when refactoring the test folder?

Speaking of which, I had a look at it earlier and felt so lost I didn't get anywhere... I kind of gave up for now and returned to other issues :/

Reply:
Yeah, just forget about the test folder stuff for now. I had a look too. It's a bit of a mess.

We define tests with .yaml, and link to a bunch of stuff via that, but then we expect .rst and .stderr to be there with the same basename. Arguably everything should be referenced via .yaml, and not magically assumed to be there with the same basename. It's nice to move stuff around when you don't have to worry about implicit stuff like that. OTOH that would be pretty verbose in the .yaml. Dunno.

Reply:
As to the Clang version issue, in this case the right fix might be to just fix the issue that is being warned about. I general, I'd prefer having the tests warning free apart from tests that test the warnings.

Reply:
It might also be the *easiest* fix.

Reply:
> It might also be the _easiest_ fix.

Good point, should have started there now that I think about it.

> We define tests with .yaml, and link to a bunch of stuff via that, but then we expect .rst and .stderr to be there with the same basename. Arguably everything should be referenced via .yaml, and not magically assumed to be there with the same basename. It's nice to move stuff around when you don't have to worry about implicit stuff like that. OTOH that would be pretty verbose in the .yaml. Dunno.

I'd like the explicit route actually. It's the easiest to use and, factoring the duplicated rst files, it should actually yield lower verbosity overall.

Reply:
> Good point, should have started there now that I think about it.

I fiddled with it a bit, and while I like to think I'm pretty good at C, I wasn't able to come up with an initializer for this:

```
const char* (*const *volatile *restrict legal_type_involving_function_pointer[12][3])(const char *in);
```

Maybe it should be split, idk.

Reply:
Oh, I fixed that locally already. Takes a very small piece of boilerplate, it's fairly obvious once you see it I think.

I did reduce the array size so that I didn't need to create a huge 12x3 matrix. And I still changed the `restrict` to `const` here so I can use this file for C++, but the boilerplate would be very similar either way.

I created a separate 'restrict' test unit that will be C only. Since `restrict` is not standard C++, I don't want to be testing those dark compiler corners in our tests.

I'll try to push all the fixes after dinner.